build.py

```diff
@@ -63,6 +63,7 @@
 # Defaults
 arch = "arm"
 cross_compile = cross_compilers[arch]
+build_environment = None
```

Comment:
Shouldn't we be setting the default value of `build_environment` to `{cc}-{cc-version}` and use any existing `BUILD_ENVIRONMENT` environment variable to override it (line 297)?

One issue with the current implementation is that if CC and CC_VERSION aren't defined in the environment, then the build environment ends up being `None-unknown`.

So the default value should be using the local variables `cc` and `cc_version` instead, but I think it should also not be using this default value when a `BUILD_ENVIRONMENT` is defined in the environment so that it can be set to an arbitrary string (e.g. a Docker image name and fingerprint).

Reply:
Agreed, I will make it mandatory and overrideable.

Reply:
So is the jenkins job going to set BUILD_ENVIRONMENT, or CC and CC_VERSION?

Currently build.jpl uses CC and CC_VERSION and I'm creating our own string from those, which I'm happy with.

Or how about all these options :) see new patch

Reply:
I would suggest this change instead:
```
diff --git a/build.py b/build.py
index 2484be83c7ea..07092b5aa990 100755
--- a/build.py
+++ b/build.py
@@ -247,6 +247,8 @@ if cross_compile and cc == "gcc":
     cc_cmd = "%s%s --version 2>&1" %(cross_compile, cc)
 cc_version = subprocess.check_output(cc_cmd, shell=True).splitlines()[0]
 
+build_environment = cc
+
 # KBUILD_OUTPUT
 kbuild_output = kbuild_output_prefix
 if arch:
@@ -294,7 +296,9 @@ if use_environment:
     git_branch = os.environ.get('BRANCH', git_branch)
     git_describe = os.environ.get('GIT_DESCRIBE', git_describe)
     git_describe_v = os.environ.get('GIT_DESCRIBE_VERBOSE', git_describe_v)
-    build_environment = "{}-{}".format(os.environ.get('CC', build_environment), os.environ.get('CC_VERSION', 'unknown'))
+    build_environment = os.environ.get('BUILD_ENVIRONMENT', build_environment)
+
+print("Build environment: {}".format(build_environment))
 
 start_time = time.time()
```
I don't think `cc_version` is suitable to use in the build environment as it can be very long, verbose and with many characters that would need to be escaped (brackets, spaces...).  For example, when using a Bootlin toolchain as a test, it would look like this:
```
gcc-aarch64-linux-gcc.br_real (Buildroot 2018.08.1-00003-g576b333) 7.3.0
```

So either what is calling `build.py` knows what it's doing (Jenkins or manual command line) and sets `BUILD_ENVIRONMENT` to something representative, or we stick to a safe default with just the name of the compiler or even something like `unknown` as this should never happen in a production environment.

Reply:
On a 2nd thought, having a command line argument for the build environment string is probably cleaner as this doesn't need to be passed via the environment.  Using `cc` as the default would seem fine as long as it's just the name of the compiler (gcc or clang) and not an absolute path with `/` characters which would end up distorting the storage URL.

Reply:
I see there's now a `-E` command line option, looks good.